components/sync15/src/telemetry.rs

```diff
@@ -0,0 +1,645 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// Manage recording sync telemetry. Assumes some external telemetry
+// library/code which manages submitting.
+
+use std::collections::HashMap;
+use std::time;
+
+use num::Num;
+
+// We use serde to serialize to json (but we never need to deserialize)
+use serde::ser::{Serialize, Serializer};
+use serde_derive::*;
+#[cfg(test)]
+use serde_json::{self, json};
+
+use crate::error::Error;
+
+// For skip_serializing_if
+// I'm surprised I can't use Num::is_zero directly, although Option::is_none works.
+// (Num's just a trait, Option is an enum, but surely there's *some* way?)
+fn skip_if_zero<T: Num + PartialEq>(v: &T) -> bool {
+    return v.is_zero();
+}
+
+// A test helper, used by the many test modules below. Is there a better way
+// to structure tests?
+#[cfg(test)]
+fn assert_json<T: ?Sized>(v: &T, expected: serde_json::Value)
+where
+    T: serde::Serialize,
+{
+    assert_eq!(
+        serde_json::to_value(&v).expect("should get a value"),
+        expected
+    );
+}
+
+/// What we record for 'when' and 'took' in a telemetry record.
+#[derive(Debug, Serialize)]
+struct WhenTook {
+    when: f64,
+    #[serde(skip_serializing_if = "skip_if_zero")]
+    took: u64,
+}
+
+/// What we track while recording 'when' and 'took. It serializes as a WhenTook,
+/// except when .finished() hasn't been called, in which case it panics.
+#[derive(Debug)]
+enum Stopwatch {
+    Started(time::SystemTime, time::Instant),
+    Finished(WhenTook),
+}
+
+impl Stopwatch {
+    fn new() -> Self {
+        Stopwatch::Started(time::SystemTime::now(), time::Instant::now())
+    }
+
+    fn finished(&self) -> Self {
+        match cfg!(test) {
```

Comment:
A better way to do this is probably have #[cfg(test)] and #[cfg(not(test))] in front of the real/test-only implementations I think.

If I'm missing something and there's a reason to do it this way (to make sure the _finished impl is still checked for compile errors maybe?) then I'd prefer `if cfg!(test) {}` to matching on the bool (IMO it's clearer, and it also avoids a level of indentation if the cases were more complex).